subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/DefaultTaskOutputFilesRepository.java

```diff
@@ -0,0 +1,113 @@
+
+    private final static String CACHE_DISPLAY_NAME = "Build Output Cleanup Cache";
+
+    private final CacheRepository cacheRepository;
+    private final PersistentCache cacheAccess;
+    private final FileSystemMirror fileSystemMirror;
+    private final PersistentIndexedCache<String, Boolean> outputFiles; // The value is true if it is an output file, false if it is a parent of an output file
+
+    public DefaultTaskOutputFilesRepository(CacheRepository cacheRepository, Gradle gradle, FileSystemMirror fileSystemMirror, InMemoryCacheDecoratorFactory inMemoryCacheDecoratorFactory) {
+        this.cacheRepository = cacheRepository;
+        this.cacheAccess = createCache(gradle);
+        this.fileSystemMirror = fileSystemMirror;
+        this.outputFiles = cacheAccess.createCache(cacheParameters(inMemoryCacheDecoratorFactory));
+    }
+
+    private static PersistentIndexedCacheParameters<String, Boolean> cacheParameters(InMemoryCacheDecoratorFactory inMemoryCacheDecoratorFactory) {
+        return new PersistentIndexedCacheParameters<String, Boolean>("outputFiles", String.class, Boolean.class)
+            .cacheDecorator(inMemoryCacheDecoratorFactory.decorator(100000, true));
+    }
+
+    @Override
+    public boolean isGeneratedByGradle(File file) {
+        File absoluteFile = file.getAbsoluteFile();
+        return containsFilesGeneratedByGradle(absoluteFile) || isContainedInAnOutput(absoluteFile);
+    }
+
+    private Boolean isContainedInAnOutput(File file) {
+        File currentFile = file;
+        while (currentFile != null) {
+            if (outputFiles.get(currentFile.getPath()) == Boolean.TRUE) {
+                return true;
+            }
+            currentFile = currentFile.getParentFile();
+        }
+        return false;
+    }
+
+    private boolean containsFilesGeneratedByGradle(File file) {
+        return outputFiles.get(file.getPath()) != null;
+    }
+
+    @Override
+    public void recordOutputs(TaskExecution taskExecution) {
+        for (String outputFilePath : taskExecution.getDeclaredOutputFilePaths()) {
+            FileSnapshot fileSnapshot = fileSystemMirror.getFile(outputFilePath);
+            File outputFile = new File(outputFilePath);
+            boolean exists = fileSnapshot == null ? outputFile.exists() : fileSnapshot.getType() != FileType.Missing;
+            if (exists) {
+                outputFiles.put(outputFilePath, Boolean.TRUE);
+                File outputFileParent = outputFile.getParentFile();
+                while (outputFileParent != null) {
+                    String parentPath = outputFileParent.getPath();
+                    if (outputFiles.get(parentPath) != null) {
+                        break;
+                    }
+                    outputFiles.put(parentPath, Boolean.FALSE);
+                    outputFileParent = outputFileParent.getParentFile();
+                }
+            }
+        }
+    }
+
+    protected PersistentCache createCache(Gradle gradle) {
```

Comment:
This should be `private` as it is not needed from anywhere else. It should also be `static` as it is called from the constructor. That necessitates passing in the `cacheRepository`, which makes it unnecessary to store that in a field. Much cleaner altogether that way.